### PR TITLE
feat: S3 Encryption At Rest

### PR DESCRIPTION
**Pull request by aaronmell**
Added new parameters to encrypt objects at rest in s3 when configuring artifacts. Supports SSE-S3, SSE-C and SSE-KMS

Signed-off-by: Aaron Mell <aaronamell@gmail.com>

Tips:

* Maybe add you organization to [USERS.md](https://github.com/argoproj/argo-workflows/blob/master/USERS.md).
* Your PR needs to pass the required checks before it can be approved. If the check is not required (e.g. E2E tests) it does not need to pass
* Sign-off your commits to pass the DCO check: `git commit --signoff`.
* Run `make pre-commit -B` to fix codegen or lint problems. 
* Say how how you tested your changes. If you changed the UI, attach screenshots.
* If changes were requested, and you've made them, then dismis the review to get it looked at again.
* You can ask for help!

### CODE REVIEW

**Comment by aaronmell** [comment]
Needs https://github.com/argoproj/pkg/pull/46 to be merged in, and the version updated before merging. Thats why I left this as a draft. 

I ran the following tests

- With an existing bucket
  - With both serverSideCustomerKeySecret and kmsKeyId empty, verified object uploaded and was encrypted
  - With kmsKeyId set, verified object uploaded and was encrypted with kms key
  - With serverSideCustomerKeySecret set, verfied object uploaded and was encrypted
  - With both serverSideCustomerKeySecret and kmsKeyId set, got expected error message 
  - With kmsEncryptionContext set to valid json blob, workflow passed successfully
  
- With createBucketIfNotPresent set
  - With both serverSideCustomerKeySecret and kmsKeyId empty, verified bucket had encryption enabled and object uploaded and was encrypted
  - With kmsKeyId set, verified bucket encryption enabled using kms key specified and object uploaded was encrypted with key

File: pkg/apis/workflow/v1alpha1/workflow_types.go
Comment (alexec) [inline pkg/apis/workflow/v1alpha1/workflow_types.go:1937]: Should we put all encryption relation setting under a single strict named “encryption “?

Comment (aaronmell) [reply pkg/apis/workflow/v1alpha1/workflow_types.go:1937]: Like I did on the pkg? That seems easy enough.

**Review by alexec — CHANGES_REQUESTED** [review]
dismiss review when ready

File: pkg/apis/workflow/v1alpha1/workflow_types.go
Comment (aaronmell) [reply pkg/apis/workflow/v1alpha1/workflow_types.go]: Done

File: workflow/artifacts/s3/s3.go
Comment (alexec) [inline workflow/artifacts/s3/s3.go:55]: was there an `Enable` option here?

**Review by alexec — APPROVED** [review]
I'm only giving this a cursory review. It's mostly plumbing and I'm assuming you have tested.

File: workflow/artifacts/s3/s3.go
Comment (aaronmell) [reply workflow/artifacts/s3/s3.go:55]: Line 57

Comment (aaronmell) [reply workflow/artifacts/s3/s3.go:55]: Its a few lines down

**Review by alexec — APPROVED** [review]

**Pull request merged**
